core/src/main/java/com/linecorp/armeria/internal/common/stream/HttpDataDeframer.java

```diff
@@ -0,0 +1,417 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayDeque;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+import javax.annotation.Nullable;
+
+import org.reactivestreams.Processor;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.stream.DefaultStreamMessage;
+import com.linecorp.armeria.common.util.Exceptions;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.util.ReferenceCountUtil;
+import io.netty.util.concurrent.EventExecutor;
+
+/**
+ * A skeletal {@link Processor} implementation that decodes a stream of {@link HttpData}s to N objects.
+ */
+public abstract class HttpDataDeframer<T> extends DefaultStreamMessage<T> implements Processor<HttpData, T> {
```

Comment:
- How about just `HttpDeframer`?
- Would this be more useful if it's a `Processor<HttpObject, T>`?
  - Could `HttpHeaders` be handled by other `protected` callbacks? e.g. `processInformationalHeaders`, `processHeaders()`, `processTrailers`
  - This may remove the overhead of splitting a `Publisher<HttpObject>`.

Reply:
>     * Would this be more useful if it's a `Processor<HttpObject, T>`?
>       
>       * Could `HttpHeaders` be handled by other `protected` callbacks? e.g. `processInformationalHeaders`, `processHeaders()`, `processTrailers`
>       * This may remove the overhead of splitting a `Publisher<HttpObject>`.

Adding `protected` callbacks looks nice! Let me take it.